Adding Multiplayer Client Connection Status to the Viewport

AMZN-Gene opened the pull request:
## What does this PR do?
Displays the mutliplayer connection status to the viewport allowing users to see when they've connected or disconnected from the server.
Moved IMultiplayerConnectionViewportMessage so it's available outside of Editor

Fixes #9634 
Fixes #9512

## How was this PR tested?
Playing in the editor game mode and GameLauncher. Seeing that connection status was properly displayed.
![image](https://user-images.githubusercontent.com/32776221/178093272-29796679-98ad-46db-9268-864c5b70e7ad.png)


Disconnected
![image](https://user-images.githubusercontent.com/32776221/178093280-a3ad07a3-fc40-43e5-b6f5-46012626c7de.png)

junbo75 commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,146 @@
+        // Build the connection status string (just show client connected or disconnected status for now)
+        const auto multiplayerSystemComponent = AZ::Interface<IMultiplayer>::Get();
+        MultiplayerAgentType agentType = multiplayerSystemComponent->GetAgentType();        
+        if (agentType != MultiplayerAgentType::Client)
```
Curious: Do we want to display the text for ClientServer agent type?

AMZN-Gene replied:
Nothing for clientserver or dedicated server. Need to think about what that would look like... for example, showing connection status for each client? Not sure.

junbo75 commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        // Display the connection status to the top-left viewport
+        if (AzNetworking::INetworkInterface* networkInterface = AZ::Interface<AzNetworking::INetworking>::Get()->RetrieveNetworkInterface(AZ::Name(MpNetworkInterfaceName)))
```
Nit: May worth checking whether AZ::Interface<AzNetworking::INetworking>::Get() is nullptr here.

AMZN-puvvadar replied:
MP depends on AzNetworking so this _shouldn't_ be necessary. If this Get is nullptr then something has violated that dependency.

junbo75 commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,146 @@
+                AzFramework::DebugDisplayRequestBus::Broadcast(&AzFramework::DebugDisplayRequestBus::Events::SetColor, AZ::Colors::Red);
+                AzFramework::DebugDisplayRequestBus::Broadcast(&AzFramework::DebugDisplayRequestBus::Events::Draw2dTextLabel, 
+                    editorsv_connectionStatus_topLeftBorderPadding_x, editorsv_connectionStatus_topLeftBorderPadding_y, editorsv_connectionMessageFontSize, 
+                    "Multiplayer Client Disconnected!", false);
```
Nit: May worth defining a constant for the error message.

AMZN-puvvadar replied:
Personally when it comes to error messages I prefer to keep them local to the message for searchability.

junbo75 commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.h:
```diff
@@ -0,0 +1,54 @@
+
+        //! IMultiplayerConnectionViewportMessage overrides.
+        //! @{
+        void DisplayCenterViewportMessage(const char* text) override;
```
Curious: Do we want to have some console commands to enable/disable the view port messages?

AMZN-Gene replied:
Fixed! Added _cl_viewportConnectionStatus_

AMZN-puvvadar commented on Code/Framework/AzNetworking/AzNetworking/ConnectionLayer/ConnectionEnums.h:
```diff
@@ -34,6 +34,22 @@ namespace AzNetworking
         , Connecting
     );
 
+    inline const char* GetEnumString(ConnectionState connectionState)
+    {
+        switch(connectionState)
+        {
+        case ConnectionState::Disconnected:
+            return "Disconnected";
+        case ConnectionState::Disconnecting:
+            return "Disconnecting";
+        case ConnectionState::Connected:
+            return "Connected";
+        case ConnectionState::Connecting:
+            return "Connecting";
+        }
+        return "INVALID";
+    }
```
ConnectionState is an AZ_ENUM_CLASS which defines a constexpr ToString. You can just use that.

AMZN-Gene replied:
fixed!

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,146 @@
+
+namespace Multiplayer
+{
+    const float defaultConnectionMessageFontSize = 0.7f;
```
NP: This could be constexpr

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,146 @@
+                // Display the connection status (calling VisitConnections(), but there's only 1 since we're a client)
+                auto displayConnectionStatus = [](AzNetworking::IConnection& connection)
+                {
+                    AZStd::string connectionStatusText = AZStd::string::format("Operating as multiplayer Client: %s", GetEnumString(connection.GetConnectionState()));
```
```suggestion
                    AZStd::string connectionStatusText = AZStd::string::format("Operating as multiplayer Client: %s", ToString(connection.GetConnectionState()));
```

ConnectionState is an AZ_ENUM_CLASS so you can use its constexpr ToString.

AMZN-puvvadar replied:
Also, if this is strictly read only and limited to this scope (including bus usage) you could use AZStd::string_view over AZStd::string.

AMZN-Gene replied:
fixed!

kberg-amzn replied:
None of it will actually end up compile-time though. Const-expr won't offer any meaningful benefit in these cases. I would just do whatever is simplest and least likely to cause errors moving forward. Potentially non-null terminated .data() being fed into format strings is a pretty big no-no..

AMZN-puvvadar replied:
I was actually thinking there might be an extra copy here but it could be a misread on my part.

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorConnection.cpp:
```diff
@@ -238,14 +238,14 @@ namespace Multiplayer
         }
         else
         {   
             char connection_fail_message[256];
```
I believe the preference here is to use AZStd::fixed_string<256>

AMZN-Gene replied:
much cleaner :)

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,146 @@
+                // Display the connection status (calling VisitConnections(), but there's only 1 since we're a client)
+                auto displayConnectionStatus = [](AzNetworking::IConnection& connection)
+                {
+                    AZStd::string connectionStatusText = AZStd::string::format("Operating as multiplayer Client: %s", GetEnumString(connection.GetConnectionState()));
```
You still need to change GetEnumString to ToString. Not sure how this is compiling now that you've removed the GetEnumString implementation. Was the file maybe missed in your last commit?

AMZN-Gene replied:
must have been a missed commit. my code is using ToString :shrug: ha

rainbj commented:
Hi  @AMZN-Gene - Agree with the need for this messaging, however I am a bit concerned about content overlap. I assume this message is in editor view and game mode?  I suggest putting this text in a fallback location and possibly using a different color. Red for disconnect. Green for connected
![2 column layout](https://user-images.githubusercontent.com/82551958/178058712-934b5e0d-2a24-40ce-8998-9dc225ddb93e.png)

AMZN-Gene commented:
> Hi @AMZN-Gene - Agree with the need for this messaging, however I am a bit concerned about content overlap. We currently have a fair bit amount of working going to what is called the "cluster controls". These two bits might have collisions. I assume this message is in editor view and game mode? I suggest putting this text in a fallback location and possibly using a different color. Red for disconnect. Green for connected ![2 column layout](https://user-images.githubusercontent.com/82551958/178058712-934b5e0d-2a24-40ce-8998-9dc225ddb93e.png)

Fixed! Adding color and moved the text to the bottom right. Thanks :D

AMZN-Olex commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerConnectionViewportMessage.h:
```diff
@@ -0,0 +1,28 @@
+
+        //! Displays multiplayer debug messaging in the viewport.
+        //! This will appear in the center of the screen.
+        virtual void DisplayCenterViewportMessage(const char* text) = 0;
```
What about providing an offset here as an option? The center of the screen is too specific for a generic interface here, imho.

AMZN-Gene replied:
Good call! Making me think... Setting this text will override previous text, so allowing users to define the location would give the false sense that this text is independent. Maybe this class has given the wrong impression; it's not meant to be an interface for ppl to use. I should really close it from the outside world (similar to how I dont allow users to define the bottom-right connection status and instead the class generates that itself). The center text is only meant to show the editor->server steps which didn't have an ebus I could listen into so I exposed this API directly.

AMZN-Gene replied:
Created a bug to fix this; it'll take a bit of more code so will post this in a new PR. 
https://github.com/o3de/o3de/issues/10614

AMZN-Olex commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,195 @@
+    
+    MultiplayerConnectionViewportMessageSystemComponent::MultiplayerConnectionViewportMessageSystemComponent()
+    {
+        AZ::Interface<IMultiplayerConnectionViewportMessage>::Register(this);
```
You can derive from ::Registrar to achieve the same thing.

AMZN-Gene replied:
that's awesome, thank you! fixed!

AMZN-Olex commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,195 @@
+    
+    void MultiplayerConnectionViewportMessageSystemComponent::StopCenterViewportDebugMessaging()
+    {
+        m_centerViewportDebugText = "";
```
One of these days I am going to be too lazy to see what we do with this operation? Is it the same as calling .clear() or does it create temporary objects?

AMZN-Gene replied:
Definitely not the same as clear, ends up calling assign. Switched to use clear :P

AMZN-Olex commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.h:
```diff
@@ -0,0 +1,60 @@
+    private:
+        //! AZ::TickBus::Handler
+        //! @{
+        void OnTick(float, AZ::ScriptTimePoint) override;
```
You can also use AZ::ScheduledEvent.

AMZN-Gene replied:
I think OnTick make sense since it aligns will a call every time the engine renders. Updated from OnTick to OnRenderTick so it's more explicit/correct.

AMZN-Gene commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.h:
```diff
@@ -0,0 +1,60 @@
+
+        void DrawConnectionStatus(AzNetworking::ConnectionState connectionState);
+
+        AZStd::string m_centerViewportDebugText;
```
Sweet, fixed!

junbo75 commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerConnectionViewportMessageSystemComponent.cpp:
```diff
@@ -0,0 +1,186 @@
+        AZ::RPI::ViewportContextPtr viewport = AZ::RPI::ViewportContextRequests::Get()->GetDefaultViewportContext();
+        if (!viewport)
+        {
+            return;
```
Nit: Do we want to throw an error or warning in this case? Same for all the other checking below.

AMZN-Gene replied:
Fixed! Making them WarnOnce since this is happening ontick; it might be valid to simply stop rendering entirely if these checks return null, but some of these are using old cry systems so i'm not sure about their life cycle so I'm okay with just keepign a warning.

AMZN-Gene replied:
Good call out! Hmm, this is also running on ServerLauncher which wont have a renderer so logging warnings causes spam (even AZWarnOnce is getting called enough times to become noticeable, not sure why)
Also, the font renderers life cycle isn't guaranteed (looks like its using cry systems); until that changes, it might be best to leave the checks inside OnTick, but not warn in order to avoid spam :c

AMZN-Gene replied:
I found a way to clean this up so the system is better at coming alive when needed instead of waiting around. It’ll require some more callbacks for MP system, I’ll add that to a new Pr.

junbo75 reviewed (approved):
Looks good to me. Unfortunately it's hard to add tests for these UI changes.
